lib/rules/no-useless-flag.ts

```diff
@@ -0,0 +1,631 @@
+                    return false
+                }
+            }
+        }
+        // Need to report
+        return true
+    }
+
+    public pushReadNode(node: Expression) {
+        this.readNodes.set(node, {})
+    }
+
+    public setDefineId(codePathId: string, loopNode: Statement | undefined) {
+        this.defineId = { codePathId, loopNode }
+    }
+
+    public markAsUsedInSearchOrSplit(node: Expression) {
+        const exprState = this.readNodes.get(node)
+        if (exprState) {
+            exprState.marked = true
+            exprState.usedInSearchOrSplit = true
+        }
+    }
+
+    public markAsUsedInExecOrTest(
+        node: Expression,
+        codePathId: string,
+        loopNode: Statement | undefined,
+    ) {
+        const exprState = this.readNodes.get(node)
+        if (exprState) {
+            exprState.marked = true
+            exprState.usedInExecOrTest = { id: { codePathId, loopNode } }
+        }
+    }
+
+    public isUsed() {
+        return this.state.used
+    }
+
+    public markAsUsed() {
+        this.state.used = true
+    }
+
+    public markAsCannotTrack() {
+        this.state.track = false
+    }
+}
+
+/**
+ * If the given expression node is assigned with a variable declaration, it returns the variable name node.
+ */
+function getVariableId(node: Expression) {
+    const parent = getParent(node)
+    if (
+        !parent ||
+        parent.type !== "VariableDeclarator" ||
+        parent.init !== node ||
+        parent.id.type !== "Identifier"
+    ) {
+        return null
+    }
+    const decl = getParent(parent)
+    if (decl && decl.type === "VariableDeclaration" && decl.kind === "const") {
+        return parent.id
+    }
+    return null
+}
+
+/**
+ * Gets the location for reporting the flag.
+ */
+function getFlagLocation(
+    context: Rule.RuleContext,
+    node: RegExpExpression,
+    flag: "i" | "m" | "s" | "g",
+) {
+    const sourceCode = context.getSourceCode()
+    if (node.type === "Literal") {
+        const flagIndex =
+            node.range![1] -
+            node.regex.flags.length +
+            node.regex.flags.indexOf(flag)
+        return {
+            start: sourceCode.getLocFromIndex(flagIndex),
+            end: sourceCode.getLocFromIndex(flagIndex + 1),
+        }
+    }
+    return node.arguments[1].loc!
+}
+
+/**
+ * Returns a fixer that removes the given flag.
+ */
+function fixRemoveFlag(
+    fixer: Rule.RuleFixer,
+    node: RegExpExpression,
+    flag: "i" | "m" | "s" | "g",
+) {
+    if (node.type === "Literal") {
+        const flagIndex =
+            node.range![1] -
+            node.regex.flags.length +
+            node.regex.flags.indexOf(flag)
+        return [
+            // Replace the range of regular expression literals to avoid conflicts.
+            fixer.replaceTextRange([node.range![0], node.range![0] + 1], "/"),
+            fixer.removeRange([flagIndex, flagIndex + 1]),
```

Comment:
Could you please explain why this works?

From what I see, ESLint is free to apply the fixes we give independently of each other. Does ESLint guarantee that an array of fixes will always be applied atomically?

Reply:
Even if multiple ranges are given by the fixer of one report, ESLint synthesizes them and judges it as one fix. If the fix range returned by each report is duplicated, only one will be applied.

Regular expressions start with `/`, so by replacing the start position with `/`, the fix range of this report will be from the start position of the regular expression.

Reply:
https://github.com/eslint/eslint/blob/master/lib/linter/report-translator.js#L146

Reply:
The maximum number of fixes per report is one, and multiple reports are applied to the fix range without conflict in the following code:
https://github.com/eslint/eslint/blob/master/lib/linter/source-code-fixer.js#L61

Reply:
The problem that I have with this is that [ESLint doesn't guarantee this behavior](https://eslint.org/docs/developer-guide/working-with-rules#applying-fixes). ESLint is free to change this part of its implementation at any time.

I think we should not rely on this specific implementation detail.

Reply:
I'll explain this, but through this discussion I realized that this code is hard to read and requires a lot of information. I would like to change this source code.

---

I think you are worried about the possibility that the merge process (`mergeFixes`) will be removed.

I think removing `mergeFixes` involves changing the API of `LintMessage#fix`, so removing it should be a breaking change.

https://eslint.org/docs/developer-guide/nodejs-api#-lintmessage-type
https://eslint.org/docs/developer-guide/nodejs-api#-editinfo-type

Next is why `mergeFixes` is guaranteed to be done before applying the fix (`SourceCodeFixer.applyFixes`).

The `LintMessage` API can be used in the `postprocess` of `processors` in the eslint-plugin.

https://eslint.org/docs/developer-guide/working-with-plugins#processors-in-plugins

The `postprocess` is done before the `SourceCodeFixer.applyFixes`, so you can shift the fix locations. This is often done with plugins such as `eslint-plugin-markdown`, `eslint-plugin-html` and `eslint-plugin-svelte3`.

Therefore, removing `mergeFixes` is a breaking change and will not be done except for major version upgrades.

Reply:
> I think you are worried about the possibility that the merge process (mergeFixes) will be removed.
>
> I think removing mergeFixes involves changing the API of LintMessage#fix, so removing it should be a breaking change.
>
> https://eslint.org/docs/developer-guide/nodejs-api#-lintmessage-type
> https://eslint.org/docs/developer-guide/nodejs-api#-editinfo-type

That was exactly my worry. I was worried that one error with multiple fixes might be converted into many errors with one fix each.

But you are right, there is nothing to worry about. `LintMessage` provides the guarantees we need.

> The LintMessage value is the information of *each linting error*.

This is the important part. This means that every error report will result in exactly one `LintMessage`. `mergeFixes` is required to provide this guarantee.